### PR TITLE
Modular Streisand: Optional Shadowsocks

### PR DESCRIPTION
**Pull request by cpu**
This PR updates the Shadowsocks role to be optional, allowing it to be enabled/disabled by changing the `playbooks/group_vars/all` variable `streisand_shadowsocks_enabled` var. By default it is `true` and the base Streisand system remains unchanged. 

`documentation/modular_roles.md` describes the approach taken and more information can be found in #746.

Once this PR is merged I have a similar branch that makes Wireguard optional that I will submit. I'm deliberately converting the roles piece-wise and trying to keep the review pipeline relatively unclogged. Thanks!

_Edit: Apologies for the unrelated commits at the head of this PR, conflicts with master have been resolved so the diffs shown should be only from the modular Shadowsocks work. I've had this branch around for a while_

Updates https://github.com/jlund/streisand/issues/746

### CODE REVIEW

File: documentation/modular_roles.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+Similarly, initially the `streisand-mirror` role downloaded all of the
+Shadowsocks clients for the mirror using a distinct tasks `.yml` with its own
+vars `.yml`. This made it fairly easy to move the mirror subtasks from
+`streisand-mirror/tasks/shadowsocks.yml` to
+`shadowsocks/tasks/mirror.yml` and `streisand-mirror/vars/shadowsocks.yml` to
+`shadowsocks/vars/mirror.yml`. Now the `shadowsocks` role can be responsible for
+preparing its own client mirror.
+
+Gateway & Mirror Indexes
+------------------------------
```
Comment (alimakki) [inline documentation/modular_roles.md]: Indexes -> Indices?

File: documentation/modular_roles.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+While the Gateway & Mirror documentation can be largely agnostic of which
+services are included in a given build by outsourcing the management of
+subfolders of `/var/www/streisand` and `/var/www/streisand/mirror` eventually
+a consistent set of docs to navigate must be created based on the included
+services. This is done by conditionally adding links to the connection
+instruction & mirror indexes of each subservice using the enable bool vars in
```
Comment (alimakki) [inline documentation/modular_roles.md]: indexes -> indices?

Comment (alimakki) [inline documentation/modular_roles.md]: vars -> variables
(again, your call, not trying to bike-shed)

**Comment by cpu** [comment]
> This should probably be a Phase Two thing, but one gap to consider is that we're still opening up the Shadowsocks port on the Azure/EC2/GCE side. I don't really have a super strong opinion on this, but it seems like it would be cleaner to continue to keep those tasks in separate roles (e.g. azure-security-group) and add some conditionals in those focused roles rather than bringing provider-specific logic (and lots of associated conditionals) into every role. What do you think?

Oops! I think I missed this chunk on this branch and fixed it in my wireguard branch. I intended to have conditionals here for the Shadowsocks port. I agree with your rationale that its best to suffer the conditionals in the provider specific firewall roles than try to move provider specific networking into each of the service roles. Unfortunately I couldn't think of something more clever like with dnsmasq.

**Comment by cpu** [comment]
@alimakki @jlund Thank you for the detailed reviews! I believe all of your feedback has been resolved by fe4480a.

**Pull request merged**